tests/unit/optimiser/test_team_selector.py

```diff
@@ -0,0 +1,489 @@
+import numpy as np
+import pytest
+
+from footbot.optimiser.team_selector import construct_player_position_weights
+from footbot.optimiser.team_selector import construct_player_team_weights
+from footbot.optimiser.team_selector import get_elements_from_vector
+from footbot.optimiser.team_selector import get_sorted_safe_web_names
+from footbot.optimiser.team_selector import select_team
+from footbot.optimiser.team_selector import validate_optimisation_arguments
+
+
+def test_get_elements_from_vector():
+    vector = [0.0, 1.0]
+    player_elements = [1, 2]
+
+    assert get_elements_from_vector(vector, player_elements) == [2]
+
+    with pytest.raises(Exception) as e:
+        get_elements_from_vector([0.1], player_elements)
+    assert "Vector contains values other than 0. and 1." in str(e.value)
+
+    with pytest.raises(Exception) as e:
+        get_elements_from_vector([0.0], player_elements)
+    assert "`vector` and `player_elements` are different lengths" in str(e.value)
+
+
+def test_validate_optimisation_arguments():
+
+    kwargs = dict(
+        total_budget=1000,
+        first_team_factor=0.9,
+        bench_factor=0.1,
+        captain_factor=0.9,
+        vice_factor=0.1,
+        existing_squad=[],
+        transfer_penalty=4,
+        transfer_limit=15,
+    )
+
+    with pytest.raises(Exception) as e:
+        validate_optimisation_arguments(**{**kwargs, "total_budget": -1})
+    assert "`total_budget` must be positive" in str(e.value)
+
+    with pytest.raises(Exception) as e:
+        validate_optimisation_arguments(**{**kwargs, "bench_factor": 0.11})
+    assert "`first_team_factor` and `bench_factor` must sum to 1." in str(e.value)
+
+    with pytest.raises(Exception) as e:
+        validate_optimisation_arguments(**{**kwargs, "vice_factor": 0.11})
+    assert "`captain_factor` and `vice_factor` must sum to 1." in str(e.value)
+
+    with pytest.raises(Exception) as e:
+        validate_optimisation_arguments(**{**kwargs, "existing_squad": [1]})
+    assert "`existing_squad` must consist of 15 players or be an empty array" in str(
+        e.value
+    )
+
+    with pytest.raises(Exception) as e:
+        validate_optimisation_arguments(**{**kwargs, "existing_squad": False})
+    assert "`existing_squad` must consist of 15 players or be an empty array" in str(
+        e.value
+    )
+
+    with pytest.raises(Exception) as e:
+        validate_optimisation_arguments(**{**kwargs, "transfer_penalty": -1})
+    assert "`transfer_penalty` must be non-negative" in str(e.value)
+
+    with pytest.raises(Exception) as e:
+        validate_optimisation_arguments(**{**kwargs, "transfer_limit": 16})
+    assert "`transfer_limit` must be an integer between 0 and 15" in str(e.value)
+
+
+keepers = [
```

Comment:
it's idiomatic to use pytest fixtures for this sort of stuff instead of module-level variables. two benefits i can think of-

- it makes your tests remain functionally pure (no dependency on anything but their arguments and function under test)
- you can remove risk of mutating the test data in some other test and having wrong test fail

Reply:
Sounds sensible